### PR TITLE
Make outlives::{components,verify} agree

### LINKED ISSUE #97405: Outlives bounds for projections are broken when used with generators/closures
I tried this code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=776e25cac316cad8c1ee83508ceb7f69

```rust
fn opaque<F>(_: F) -> impl Iterator { b"".iter() }
fn assert_static<T: 'static>(_: T) {}

fn generic_fn<T>() {
    // proving `<OpaqueTy<type_of(async {})> as Iterator>::Item: 'static`
    // somehow requires `T: 'static`.
    assert_static(opaque(async {}).next());
    //~^ the associated type `<impl Iterator as Iterator>::Item` may not live long enough
    assert_static(opaque(|| {}).next());
    //~^ the associated type `<impl Iterator as Iterator>::Item` may not live long enough
}
```

Both asserts don't pass. Adding `<T: 'static>` bound on `generic_fn` makes it compile.

We are requiring *all* the `Substs` for closures and generators to outlive `'static` here:
https://github.com/rust-lang/rust/blob/9fed13030c2a2ebd79bfb1fd8be4f768cbe8c9d9/compiler/rustc_infer/src/infer/outlives/verify.rs#L77

This is wrong and inconsistent with the logic in [fn compute_components<'tcx>(](https://github.com/rust-lang/rust/blob/9fed13030c2a2ebd79bfb1fd8be4f768cbe8c9d9/compiler/rustc_infer/src/infer/outlives/components.rs#L62)
### Meta

Build using the Nightly version: 1.63.0-nightly

(2022-05-24 76761db5919b6d2e178b)

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

@rustbot label T-compiler A-lifetimes A-closures A-generators
@rustbot claim

### PR DESCRIPTION
**Pull request by aliemjay**
fixes #97405

cc @oli-obk this is should fix https://github.com/rust-lang/rust/pull/95474#discussion_r880660075

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
Nominating for T-types discussion:

```rust
fn opaque<F>(_: F) -> impl Iterator { b"".iter() }
fn assert_static<T: 'static>(_: T) {}

fn good_generic_fn<T>() {
    // proving `<OpaqueTy<type_of(async {})> as Iterator>::Item: 'static`
    // somehow requires `T: 'static`.
    assert_static(opaque(async {}).next());
    assert_static(opaque(|| {}).next());
    assert_static(opaque(opaque(async {}).next()).next());
}
```

does not compile right now, because the `T` of `good_generic_fn` is inherited by generators and closures, even if they do not capture any types that relate to `T`. This PR does not change that, but it instead adjusts the algorithm figuring out which types and lifetimes must be considered for an outlives bound to only look at captured things in the generators or closures.

This scheme already works on stable for simpler cases, but in associated types it breaks down. Thus I nominate for T-types instead of T-lang, since imo this is just-an-impl-bug and the behaviour intended by the lang team.

File: compiler/rustc_infer/src/infer/outlives/components.rs [hunk 1]
```diff
@@ -190,7 +190,11 @@ fn compute_components<'tcx>(
         }
 }
 
-fn compute_components_recursive<'tcx>(
+/// Collect [Component]s for *all* the substs of `parent`.
+///
+/// This should not be used to get the components of `parent` itself.
+/// Use [push_outlives_components] instead.
+pub(crate) fn compute_components_recursive<'tcx>(
```
Comment (oli-obk) [inline compiler/rustc_infer/src/infer/outlives/components.rs:197]:
```suggestion
pub(super) fn compute_components_recursive<'tcx>(
```

File: compiler/rustc_infer/src/infer/outlives/verify.rs [hunk 2]
```diff
@@ -177,26 +145,43 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
...
+                    // ignore this, we presume it will yield an error
+                    // later, since if a type variable is not resolved by
+                    // this point it never will be
+                    self.tcx.sess.delay_span_bug(
+                        rustc_span::DUMMY_SP,
+                        &format!("unresolved inference variable in outlives: {:?}", v),
+                    );
+                    // add a bound that never holds
+                    VerifyBound::AnyBound(vec![])
                 }
```
Comment (oli-obk) [inline compiler/rustc_infer/src/infer/outlives/verify.rs:184]: Please pull this match out into a separate method

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by aliemjay** [comment]
a less contrived example that should be fixed now:
```rust
fn from_fn<F>(_: F) -> impl IntoIterator<Item = u8> { [] }

fn generic_fn<T>() -> Box<dyn Iterator<Item = u8>> {
    Box::new(from_fn(|| {}).into_iter())
    //~^ ERROR the associated type `<impl IntoIterator<Item = u8> as IntoIterator>::IntoIter` may not live long enough
}
```

**Comment by oli-obk** [comment]
Let's see if this opens a zulip thread automagically

**Comment by aliemjay** [comment]
oli-obk is in vacation, so
r? @nikomatsakis 

@nikomatsakis  choosed you because you recently touched `VerifyBound`s and  types tests, and this PR basically fixes malformed verify bounds for projection tys. Feel free to reassign if you like.

**Comment by oli-obk** [comment]
@rfcbot merge

```rust
fn opaque<F>(_: F) -> impl Iterator { b"".iter() }
fn assert_static<T: 'static>(_: T) {}

fn good_generic_fn<T>() {
    // proving `<OpaqueTy<type_of(async {})> as Iterator>::Item: 'static`
    // somehow requires `T: 'static`.
    assert_static(opaque(async {}).next());
    assert_static(opaque(|| {}).next());
    assert_static(opaque(opaque(async {}).next()).next());
}
```

does not compile right now, because the `T` of `good_generic_fn` is inherited by generators and closures, even if they do not capture any types that relate to `T`.

This PR does not change that, but it instead adjusts the algorithm figuring out which types and lifetimes must be considered for an outlives bound to only look at captured things in the generators or closures.

This scheme already works on stable for simpler cases, but in associated types it breaks down. Thus I nominate for T-types instead of T-lang, since imo this is just-an-impl-bug and the behaviour intended by the lang team.

**Comment by aliemjay** [comment]
I have rebased to adjust for NLL diagnostic regressions. The first commit has not changed.

r? @oli-obk

**Comment by jackh726** [comment]
Spurious?

@bors retry

**Pull request merged**